lib/tasks/migrate.rake

```diff
@@ -1,76 +1,226 @@
+      raise("Interrupted") unless reports_migrate_running
+      counter += 1
+      puts("Processing report #{counter} out of #{report_count} reports") if (counter % print_each).zero?
+      if r.origin == "Puppet"
+        logs = create_puppet_logs(r.id, r.logs)
+        keywords = detect_puppet_keywords(r.status, logs)
+      elsif r.origin == "Ansible"
+        results = create_ansible_results(r.id, r.logs)
+        keywords = detect_ansible_keywords(r.status)
+      end
+      if keywords.present?
+        keywords_to_insert = keywords.each_with_object([]) do |n, ks|
+          ks << { name: n }
+        end
+        ReportKeyword.upsert_all(keywords_to_insert, unique_by: :name)
+        report_keyword_ids = ReportKeyword.where(name: keywords).distinct.pluck(:id)
+      end
+      summary = summary(r.origin, r.metrics, r.status)
+      body = create_body(r.origin&.downcase, r.metrics, r.reported_at, r.status, r.host, report_keyword_ids, summary)
+      if r.origin == "Puppet"
+        body.merge!(logs: logs)
+      elsif r.origin == "Ansible"
+        body.merge!(results: results)
       end
-      body = create_body(r.metrics, r.reported_at, r.status, r.logs, r.host)
       origin = r.origin.downcase
-      host_report = host_report_create(r.host_id, origin, body)
       User.without_auditing do
         User.as_anonymous_admin do
-          HostReport.create(host_report)
+          HostReport.create!(build_report(r.host_id, origin, body, report_keyword_ids))
         end
       end
     rescue StandardError => e
-      puts "An error occurred, to resume the process: foreman-rake host_reports:migrate from=#{r.reported_at} start=#{r.id}"
-      throw e
+      puts "Error when processing report ID=#{r.id}"
+      puts r.inspect
+      puts "To resume the process:\n\n***\n\nforeman-rake host_reports:migrate from_id=#{r.id} from_date=#{from_date}\n\n***\n\n"
+      raise e
+    end
+    puts "\n\nALL DONE!\n\nCheck the migrated reports in Monitor - Host Reports first"
+    puts "and when ready, expire old configuration reports with:\n\n"
+    puts "  rake reports:expire report_type=config_report days=0\n\n"
+    puts "Report expiration is slow, if you don't use OpenSCAP plugin, then"
+    puts "truncate the following tables in the foreman database"
+    puts "for a quick delete (this will remove also OpenSCAP reports):\n\n"
+    puts "  truncate logs, messages, resources, reports;\n\n"
+    puts "Reclaim postgres database space via VACUUM function in any case."
+    puts "If migration was not successful, truncate tables host_reports and"
+    puts "report_keywords and start over.\n"
+    puts "Optionally, refresh host statuses with:\n\n"
+    puts "  foreman-rake host_reports:refresh\n\n"
+  end
+
+  desc <<-END_DESC
+  Host status information can be incorrect until new report is received.
+  This task refreshes all host statuses and global statuses.
+  END_DESC
+  task :refresh => :environment do
```

Comment:
Mabye something like `refresh_statuses` or even `refresh_host_statuses` would be a better name? `host_reports:refresh` would suggest that it refreshes reports